packages/sdk-core/src/modules/database/BacktraceDatabase.ts

```diff
@@ -0,0 +1,179 @@
+     */
+    public add(
+        backtraceData: BacktraceData,
+        attachments: BacktraceAttachment<unknown>[],
+    ): BacktraceDatabaseRecord | undefined {
+        if (!this._enabled) {
+            return undefined;
+        }
+
+        this.prepareDatabase();
+
+        const record = {
+            count: 1,
+            data: backtraceData,
+            hash: '',
+            id: IdGenerator.uuid(),
+            locked: false,
+            attachments: attachments,
+        };
+        this._storageProvider.add(record);
+        this._databaseRecordContext.add(record);
+
+        return record;
+    }
+
+    /**
+     * Returns stored references to Backtrace records
+     * @returns available records in the database
+     */
+    public get(): BacktraceDatabaseRecord[] {
+        return this._databaseRecordContext.get();
+    }
+
+    /**
+     * @returns Returns number of records stored in the Database
+     */
+    public count(): number {
+        return this._databaseRecordContext.count();
+    }
+
+    /**
+     * Disables database integration. After running the dispose method, you cannot
+     * execute any database operations.
+     */
+    public dispose() {
+        this._enabled = false;
+        clearInterval(this._intervalId);
+    }
+
+    /**
+     * Removes the database record
+     * @param record database records
+     */
+    public async remove(record: BacktraceDatabaseRecord) {
+        if (!this._enabled) {
+            return;
+        }
+        this._databaseRecordContext.remove(record);
+        await this._storageProvider.delete(record);
+    }
+
+    /**
+     * Prepare database to insert records
+     * @param totalNumberOfRecords number of records to insert
+     */
+    private prepareDatabase(totalNumberOfRecords = 1) {
+        const numberOfRecords = this.count();
+        if (numberOfRecords + totalNumberOfRecords > this._maximumRecords) {
+            this._databaseRecordContext.dropOverflow(totalNumberOfRecords);
+        }
+    }
+
+    private async loadReports(): Promise<void> {
+        const records = await this._storageProvider.get();
+        if (records.length > this._maximumRecords) {
+            records.length = this._maximumRecords;
+        }
+        this.prepareDatabase(records.length);
+        this._databaseRecordContext.load(records);
+    }
+
+    private async setupDatabaseAutoSend() {
+        if (this._options?.autoSend === false) {
+            return;
+        }
+
+        const sendDatabaseReports = async () => {
+            await this.sendRecords();
+        };
+        this._intervalId = setInterval(sendDatabaseReports, this._retryInterval);
+        await this.sendRecords();
+    }
+
+    private async sendRecords() {
+        for (let bucketIndex = 0; bucketIndex < this._databaseRecordContext.bucketCount; bucketIndex++) {
+            for (const record of this._databaseRecordContext.getBucket(bucketIndex)) {
+                if (record.locked) {
```

Comment:
I'm a little confused by the purpose of the locked logic. Would you mind explaining?

Reply:
Ah that makes sense. Thank you!